homeassistant/components/media_player/ps4.py

```diff
@@ -0,0 +1,510 @@
+
+    @util.Throttle(MIN_TIME_BETWEEN_SCANS, MIN_TIME_BETWEEN_FORCED_SCANS)
+    def update(self):
+        """Retrieve the latest data."""
+        try:
+            status = self._ps4.get_status()
+            state = status.get('status')
+            if state == 'Ok':
+                titleid = status.get('running-app-titleid')
+                name = status.get('running-app-name')
+                if titleid and name is not None:
+                    self._state = STATE_PLAYING
+                    self._media_content_id = titleid
+                    app_name = self.get_ps_store_data('title', name)
+                    if app_name is None:
+                        app_name = name
+                    self._media_image = self.get_ps_store_data('art', app_name)
+                    self._source = app_name
+                    self._media_title = app_name
+                    if titleid in self._games:
+                        store = self._games[titleid]
+                        if store != app_name:
+                            self._games.pop(titleid)
+                    if titleid not in self._games:
+                        self.add_games(titleid, app_name)
+                        self._games = self.load_games()
+                    self._source_list = list(sorted(self._games.values()))
+                else:
+                    self.idle()
+                    if self._source_selected is not None:
+                        self._source_selected = None
+            else:
+                self.state_off()
+        except socket.timeout:
+            pass
+
+    def idle(self):
+        """Set states for state idle."""
+        self.no_title()
+        self._state = STATE_IDLE
+
+    def state_off(self):
+        """Set states for state off."""
+        self.no_title()
+        self._state = STATE_OFF
+
+    def no_title(self):
+        """Update if there is no title."""
+        self._media_title = None
+        self._media_content_id = None
+        self._source = None
+        self._source_selected = None
+
+    def load_games(self):
+        """Load games for sources."""
+        file = self._games_filename
+        try:
+            games = load_json(file)
+            return games
+        except FileNotFoundError:
+            games = {}
+            self.save_games(games)
+            self.load_games()
+
+    def save_games(self, games):
+        """Save games to file."""
+        file = self._games_filename
+        try:
+            save_json(file, games)
+        except OSError as error:
+            _LOGGER.error("Could not save game list, %s", error)
+
+    def add_games(self, titleid, app_name):
+        """Add games to list."""
+        games = self.load_games()
+        if titleid is not None and titleid not in games:
+            game = {titleid: app_name}
+            games.update(game)
+            self.save_games(games)
+
+    def get_ps_store_data(self, data_type, title):
```

Comment:
This should be moved to a standalone library published on pypi.

Reply:
The function at 242? Please elaborate.

Reply:
Transfered both to imported package.